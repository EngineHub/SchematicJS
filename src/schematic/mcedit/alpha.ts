import type { TagMap, Int } from 'nbt-ts';
import type { Block } from '../types.js';
import { Schematic } from '../types.js';
import legacyMapper from './legacy.json';

export function loadMCEditAlpha(tag: TagMap): Schematic {
    const materials = tag.get('Materials') as string;
    const blockIds = tag.get('Blocks') as Buffer;
    const blockData = tag.get('Data') as Buffer;
    const width = (tag.get('Width') as Int).value;
    const height = (tag.get('Height') as Int).value;
    const length = (tag.get('Length') as Int).value;

    const palette = new Map<string, Block>();

    for (let x = 0; x < width; ++x) {
        for (let y = 0; y < height; ++y) {
            for (let z = 0; z < length; ++z) {
                const index = y * width * length + z * width + x;
                const id = `${blockIds[index]}:${blockData[index]}`;
                let key = legacyMapper.blocks[id];
                if (!key) {
                    throw new Error('Unknown legacy block: ' + id);
                }

                // sanitize the block name
                const colonIndex = key.indexOf(':');
                if (colonIndex !== -1) {
                    key = key.substring(colonIndex + 1);
                }

                const properties = {};

                const bracketIndex = key.indexOf('[');
                let type: string;
                if (bracketIndex !== -1) {
                    type = key.substring(0, bracketIndex);
                    const propertyArea = key.substring(
                        bracketIndex + 1,
                        key.length - 1
                    ) as string;
                    propertyArea.split(',').forEach(prop => {
                        const pair = prop.split('=');
                        properties[pair[0]] = pair[1];
                    });
                } else {
                    type = key;
                }

                palette.set(id, { type, properties });
            }
        }
    }

    const schematic = new Schematic({
        width,
        height,
        length,
        blockTypes: [...palette.values()],
        dataVersion: undefined,
        metadata: {},
        format: {
            type: 'mcedit',
            version: materials === 'Alpha' ? 1 : undefined
        }
    });

    for (let x = 0; x < width; ++x) {
        for (let y = 0; y < height; ++y) {
            for (let z = 0; z < length; ++z) {
                const index = y * width * length + z * width + x;
                const id = `${blockIds[index]}:${blockData[index]}`;
                const block = palette.get(id);
                schematic.setBlock({ x, y, z }, block);
            }
        }
    }

    return schematic;
}
