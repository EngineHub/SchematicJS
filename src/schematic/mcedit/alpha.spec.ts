import fs from 'fs';
import type { TagMap } from '@enginehub/nbt-ts';
import { decode } from '@enginehub/nbt-ts';
import { loadMCEditAlpha } from './alpha';
import { unzip } from 'gzip-js';

function parseNbt(nbt: Uint8Array): TagMap {
    const deflated = unzip(nbt);
    const data = decode(Buffer.from(deflated), {
        unnamed: false,
        useMaps: true
    });
    return data.value as TagMap;
}

describe('loadMCEditAlpha', () => {
    it('should load a MCEdit Alpha schematic', () => {
        const loadedFile = new Uint8Array(
            fs.readFileSync('./src/test-files/mcedit_alpha.schematic')
        );
        const schematic = loadMCEditAlpha(parseNbt(loadedFile));
        expect(schematic).toMatchSnapshot();
    });
});
