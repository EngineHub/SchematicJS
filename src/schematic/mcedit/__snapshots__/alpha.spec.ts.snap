// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadMCEditAlpha should load a MCEdit Alpha schematic 1`] = `
Schematic {
  "blockTypes": [
    {
      "properties": {},
      "type": "stone",
    },
    {
      "properties": {
        "delay": "1",
        "facing": "west",
        "locked": "false",
        "powered": "false",
      },
      "type": "repeater",
    },
    {
      "properties": {},
      "type": "air",
    },
    {
      "properties": {},
      "type": "stone_bricks",
    },
    {
      "properties": {
        "extended": "false",
        "facing": "east",
      },
      "type": "piston",
    },
    {
      "properties": {},
      "type": "glass",
    },
    {
      "properties": {
        "rotation": "11",
      },
      "type": "sign",
    },
    {
      "properties": {
        "delay": "4",
        "facing": "north",
        "locked": "false",
        "powered": "false",
      },
      "type": "repeater",
    },
    {
      "properties": {
        "east": "none",
        "north": "none",
        "power": "0",
        "south": "none",
        "west": "none",
      },
      "type": "redstone_wire",
    },
    {
      "properties": {
        "has_record": "false",
      },
      "type": "jukebox",
    },
    {
      "properties": {
        "extended": "false",
        "facing": "east",
      },
      "type": "sticky_piston",
    },
    {
      "properties": {
        "delay": "4",
        "facing": "south",
        "locked": "false",
        "powered": "false",
      },
      "type": "repeater",
    },
    {
      "properties": {
        "rotation": "15",
      },
      "type": "sign",
    },
    {
      "properties": {
        "delay": "1",
        "facing": "north",
        "locked": "false",
        "powered": "false",
      },
      "type": "repeater",
    },
    {
      "properties": {
        "extended": "false",
        "facing": "west",
      },
      "type": "sticky_piston",
    },
    {
      "properties": {
        "facing": "south",
        "type": "single",
      },
      "type": "chest",
    },
    {
      "properties": {
        "rotation": "0",
      },
      "type": "sign",
    },
    {
      "properties": {
        "enabled": "false",
        "facing": "south",
      },
      "type": "hopper",
    },
    {
      "properties": {
        "facing": "west",
        "triggered": "true",
      },
      "type": "dropper",
    },
    {
      "properties": {
        "delay": "1",
        "facing": "south",
        "locked": "false",
        "powered": "false",
      },
      "type": "repeater",
    },
    {
      "properties": {
        "extended": "false",
        "facing": "south",
      },
      "type": "piston",
    },
    {
      "properties": {
        "facing": "north",
      },
      "type": "wall_torch",
    },
    {
      "properties": {
        "rotation": "8",
      },
      "type": "sign",
    },
  ],
  "blocks": [
    [
      [
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
      ],
      [
        {
          "properties": {
            "delay": "1",
            "facing": "west",
            "locked": "false",
            "powered": "false",
          },
          "type": "repeater",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "extended": "false",
            "facing": "east",
          },
          "type": "piston",
        },
        {
          "properties": {},
          "type": "glass",
        },
        {
          "properties": {},
          "type": "glass",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "rotation": "11",
          },
          "type": "sign",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
    ],
    [
      [
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
      ],
      [
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {
            "delay": "4",
            "facing": "north",
            "locked": "false",
            "powered": "false",
          },
          "type": "repeater",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
      ],
      [
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "has_record": "false",
          },
          "type": "jukebox",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "extended": "false",
            "facing": "east",
          },
          "type": "sticky_piston",
        },
        {
          "properties": {
            "delay": "4",
            "facing": "south",
            "locked": "false",
            "powered": "false",
          },
          "type": "repeater",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "rotation": "15",
          },
          "type": "sign",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
    ],
    [
      [
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone",
        },
      ],
      [
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
      ],
      [
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {
            "delay": "1",
            "facing": "north",
            "locked": "false",
            "powered": "false",
          },
          "type": "repeater",
        },
        {
          "properties": {
            "extended": "false",
            "facing": "west",
          },
          "type": "sticky_piston",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "facing": "south",
            "type": "single",
          },
          "type": "chest",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "rotation": "0",
          },
          "type": "sign",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
    ],
    [
      [
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
        {
          "properties": {},
          "type": "stone",
        },
      ],
      [
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
        {
          "properties": {},
          "type": "stone_bricks",
        },
      ],
      [
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "glass",
        },
        {
          "properties": {
            "enabled": "false",
            "facing": "south",
          },
          "type": "hopper",
        },
        {
          "properties": {
            "facing": "west",
            "triggered": "true",
          },
          "type": "dropper",
        },
        {
          "properties": {
            "delay": "1",
            "facing": "south",
            "locked": "false",
            "powered": "false",
          },
          "type": "repeater",
        },
        {
          "properties": {
            "east": "none",
            "north": "none",
            "power": "0",
            "south": "none",
            "west": "none",
          },
          "type": "redstone_wire",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "extended": "false",
            "facing": "south",
          },
          "type": "piston",
        },
        {
          "properties": {
            "facing": "south",
            "type": "single",
          },
          "type": "chest",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "facing": "north",
          },
          "type": "wall_torch",
        },
      ],
      [
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {
            "rotation": "8",
          },
          "type": "sign",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
        {
          "properties": {},
          "type": "air",
        },
      ],
    ],
  ],
  "dataVersion": undefined,
  "format": {
    "type": "mcedit",
    "version": 1,
  },
  "height": 5,
  "length": 7,
  "metadata": {},
  "width": 4,
}
`;
